doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -1,66 +1,165 @@
+   let t_H := type of H in idtac "type of H:" t_H;
+   let tt_H := type of t_H in idtac "type of (type of H):" tt_H.
+
+A proof script, consisting of the tactics that are applied to prove a
+theorem, is often informally referred to as a "proof".
+The real proof, whether complete or incomplete, is a term, the :gdef:`proof term`,
+which users may occasionally want to examine.  (This is based on the
+*Curry-Howard isomorphism* :cite:`How80,Bar81,Gir89,H89`, which is
+a correspondence between between proofs and terms of λ-calculus.)
+
+The :cmd:`Show Proof` command displays the incomplete proof term
+before you've completed the proof.  For example:
+
+.. coqtop:: all
+
+   Show Proof.
+
+The incomplete parts, the goals, are represented by existential variables
+with names that begin with `?Goal`.  In the :cmd:`Show Existentials` command,
+each existential shows the hypotheses and conclusion for the associated goal.
+:gdef:`Existential variables <existential variable>` represent as yet unknown
+values.
+
+.. coqtop:: all
+
+   Show Existentials.
+
+Coq's kernel verifies the correctness of proof terms when it exits
+proof mode by checking that the theorem statement is the same as the
+type of the proof term.
+
+After a proof is completed, :cmd:`Print` `<theorem_name>`
+shows the proof term and its type.  The type appears after
+the colon, as for this theorem from Coq's standard library:
+
+.. coqtop:: all
+
+   Print proj1.
 
 .. _proof-editing-mode:
 
-Entering and leaving proof editing mode
----------------------------------------
+Entering and exiting proof mode
+-------------------------------
+
+Coq enters *proof mode* when you begin a proof through
+commands such as :cmd:`Theorem` or :cmd:`Goal`.  Coq user interfaces
+usually have a way to indicate that you're in proof mode.
+For example, CoqIDE shows the name of the theorem being proved at the
+bottom left of the window.  Coqtop changes its input prompt to
+include the name of the theorem being proven.
+
+.. todo PR: the last 3 sentences above seem pretty unimportant
+   don't you think the user will figure this out on their own?
+
+Tactics are only available only in proof mode, while commands
+are always available.  By convention, tactics begin with lowercase letters
+while commands begin with uppercase letters.
+
+.. todo PR: is the following really true?  I see that Focus is accepted outside a proof:
+
+   In Coq’s proof mode ... user has access to specialized
+   commands dealing with proof development pragmas documented in this
+   section.
 
-The proof editing mode is entered by asserting a statement, which typically is
-the assertion of a theorem using an assertion command like :cmd:`Theorem`. The
-list of assertion commands is given in :ref:`Assertions`. The command
-:cmd:`Goal` can also be used.
+.. todo PR: following seems to conflate the proof context with section-level info, e.g.
+   Parameter, ... are part of the section  Is this truly the "local context"?
```

Comment:
Ah, clear as mud.  Replacing `Qed` with the following, I don't know what to make of the output.

It's 3 levels, right? "local context" means things that are not in the proof state, i.e. section-scoped items?  The scope of a section can't go beyond a single file?  `Require Import` doesn't load sections although sections can be used to prevent exporting items?  Is that right?

I think we need a paragraph that defines these scopes.  Not sure if it should be here, near the `local` attribute or near the `Section` command.

```
Save bar.
Print bar.

bar = p
     : P
```

Reply:
BTW, I see we use the term "environment" 109 times.  That's usually the same as the "global context", right?  Should we replace those uses with "global context"?

Reply:
>It's 3 levels, right?

Correct. I would call these levels "global context", "section context" and "proof context". "local context" encompasses the latter two.

>The scope of a section can't go beyond a single file?

Correct but probably irrelevant to the current discussion.

>sections can be used to prevent exporting items?

Some stuff is not exported outside sections, but it's more of an implementation limitation than their actual use case. The use case of sections is to not repeat some variables and hypotheses that you would like to use in several lemmas or definitions. Therefore, it is expected that when you define a section variable or hypothesis it will be used almost the same as when you define those in your lemma statement.

Reply:
> BTW, I see we use the term "environment" 109 times. That's usually the same as the "global context", right? Should we replace those uses with "global context"?

I see many "local environment" and "global environment". Yes, it means the same as "local context" and "global context". Sometimes, "environment" is use unqualified to designate the union of the two. There are also a lot of uses of "environment variable" that have nothing to do with this.

Reply:
I wasn't so sure about changing some of the uses of "environment" to "global context", particularly in, say, the `cic` chapter.  I made those changes in a separate commit.  Let me know what you think.  If they need to be handled on a case by case basis, perhaps you could push a commit that uses the term in the right places?

I don't think "section context" is needed.  I probably skipped adjusting some of the tactics and SSR chapters.  WDYT?

Reply:
Regarding uniformizing the terminology to one of "global environment" or "global context", the two should really mean the same thing, so if we want to commit to this uniformization, there is no strict reason to do it everywhere but in the CIC (typing rules) chapter. Now, the fact that [this chapter clearly defines both "local context" and "global environment"](https://coq.github.io/doc/master/refman/language/cic.html#id6) (clearly refusing to use a symmetric terminology) makes me think that we should accept this and use those terms everywhere.

Reply:
BTW, these paragraphs also use interesting terminology: "declaration" for what you called "fact" (can be local or global) and they separate these declarations into "assumptions" on the one hand and "definitions" on the other hand. They use the term "variables" to designate all the names (be it for assumptions and definitions) and do not make a special case for hypotheses (although IMHO it's still worth doing to explain the difference in the auto-generated names).

Reply:
> if we want to commit to this uniformization, there is no strict reason to do it everywhere but in the CIC (typing rules) chapter.

I wasn't sure whether it was being used in a different sense, the CIC chapter being more math than software.  I thought I'd ask before changing them.  If there are published papers that use these terms, that would argue to keep these terms.  Also cic uses `E` for the environment (273 uses of "E" across all the doc, some/many not referring to the environment)--changing that might confusing to those who know the theory.  I see about 20 uses of "global environment", plus at least as many places where "environment" alone appears to have the same meaning.

> Now, the fact that this chapter clearly defines both "local context" and "global environment" (clearly refusing to use a symmetric terminology) makes me think that we should accept this and use those terms everywhere.

"local context" is only used 13 times in the cic chapter.  Does the definition therein "A *local context* is an ordered list of *local declarations* of names which we call *variables*." coincide with "goal context" or "local context" (section context + goal contexts) as I've defined them elsewhere?  Is the section context ordered?

I think the symmetry (actually consistency or the use of a common root term) is a plus; the local context and the global context hold things that are pretty similar.  Certainly I examine the wording that's already being used, but that's only one factor.  "global environment" seems wordy; there is no "local environment".  "environment" by itself is simpler.  If we think that changing "E" in the equations to "G" or "GC" is unwise I'd be OK with "environment" and "local context" with only a little reluctance.

BTW, the common term is "standardization" rather than "uniformization".

Reply:
> Does the definition therein "A _local context_ is an ordered list of _local declarations_ of names which we call _variables_." coincide with "goal context" or "local context" (section context + goal contexts) as I've defined them elsewhere? Is the section context ordered?

Yes to both.

I think we clearly need to involve other people in this discussion on terminology. For instance, I wouldn't be able to answer to the question about the standard terminology in published papers. Therefore, we should probably delay any further change beyond this PR (because it would otherwise slow this PR, that is almost ready, too much). The current state with just a few renamings is probably fine.

Reply:
> Yes to both.

The first question is an either-or; "yes" is not such a helpful response.  Which is it?

Reply:
Sorry, I read too fast. It coincides with "local context".